### PR TITLE
[#171643781] test suite mvp

### PR DESCRIPTION
**Pull request by balanza**
With this pr we implemented authenticated requests to the IO-app backend. It depends on this [PR](https://github.com/pagopa/io-api-testing/pull/1) so we might wait for that to be merged first.

 What's in this PR:

1. The suite accepts SPID_SESSION_TOKEN as parameter
2. The suite performs http calls to the following functionalities and checks if the result is in the expected format
    * getServices
    * getServicesByRecipient
    * getApiUserProfile
    * getUserProfile
    * getService

What's NOT in the PR:

1. test for data correctness

### CODE REVIEW

**Comment by gunzip** [comment]
in no way we can merge a PR which is marked as draft :)

File: integrations/getServicesByRecipient/index.test.ts [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+
+    expect(response.status).toBe(expectedHttpCode);
+
+    if (!decoded)
+      throw new Error(`Expected response to be in the correct format`);
+
+    decoded.fold(
+      e => {
+        console.error(e);
+        throw new Error(`Expected response to be in the correct format`);
```
Comment (gunzip) [inline integrations/getServicesByRecipient/index.test.ts]: ditto

File: integrations/getServices/index.test.ts [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+
+    expect(response.status).toBe(expectedHttpCode);
+
+    if (!decoded)
+      throw new Error(`Expected response to be in the correct format`);
+
+    decoded.fold(
+      e => {
+        console.error(e);
+        throw new Error(`Expected response to be in the correct format`);
```
Comment (gunzip) [inline integrations/getServices/index.test.ts]: do not throw, just let the test fail

File: integrations/getUserProfile/index.test.ts [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+    const decoded = await decoder(response);
+
+    expect(response.status).toBe(expectedHttpCode);
+
+    if (!decoded)
+      throw new Error(`Expected response to be in the correct format`);
+
+    decoded.fold(
+      _ => {
+        throw new Error(`Expected response to be in the correct format`);
```
Comment (gunzip) [inline integrations/getUserProfile/index.test.ts]: ditto

File: lib/env.ts [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+export const EnvMap = t.exact(
+  t.partial({
+    SPID_SESSION_TOKEN: t.string,
+    IO_BACKEND_HOST: t.string,
+    IO_BACKEND_BASEPATH: t.string
+  })
+);
+export type EnvMap = t.TypeOf<typeof EnvMap>;
+
+const pickEnvKeys = (obj: any) =>
```
Comment (gunzip) [inline lib/env.ts]: can we replace the whole methods here with a call to `EnvMap.decode(process.env)` when needed ?

File: setup/index.ts [hunk 5]
```diff
@@ -1,11 +1,55 @@
...
+  const sessionTokenTask = ask("SPID_SESSION_TOKEN", askSessionToken);
+  const backendHostTask = ask("IO_BACKEND_HOST", askIOBackendHost);
+  const backendBasepathTask = ask("IO_BACKEND_BASEPATH", askIOBackendBasePath);
+
+  const sequence = array.sequence(taskEitherSeq);
+
+  return sequence([backendHostTask, backendBasepathTask, sessionTokenTask])
+    .fold(
+      () => {
+        throw new ParamReadError();
```
Comment (gunzip) [inline setup/index.ts]: ditto

**Comment by balanza** [comment]
@gunzip right, just to note that this is child of the previous PR, which is still unresolved

> in no way we can merge a PR which is marked as draft :)

File: integrations/getServices/index.test.ts [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+      },
+      ({ value }) => {
+        expect(PaginatedServiceTupleCollection.is(value)).toBe(true);
```
Comment (gunzip) [inline integrations/getServices/index.test.ts]: when doing conditional expectation you have to assert the number of expectations needed to pass the tests

File: integrations/getUserProfile/index.test.ts [hunk 7]
```diff
@@ -0,0 +1,76 @@
...
+
+    const client = (): Promise<Response> => fetchApi(endpoint, { headers });
+
+    const response = await client();
+    const decoded = await decoder(response);
+
+    expect(response.status).toBe(expectedHttpCode);
+
+    if (!decoded)
+      throw new Error(`Expected response to be in the correct format`);
```
Comment (gunzip) [inline integrations/getUserProfile/index.test.ts]: ditto

File: setup/index.ts [hunk 8]
```diff
@@ -1,11 +1,55 @@
...
+  getEnvValue,
+  setAllEnvValues,
+  EnvMap
+} from "../lib/env";
+
+import { tryCatch, taskEitherSeq } from "fp-ts/lib/TaskEither";
+import { toError } from "fp-ts/lib/Either";
+import { array } from "fp-ts/lib/Array";
+
+const ask = (key: keyof EnvMap, promptFn: () => Promise<string>) => {
```
Comment (gunzip) [inline setup/index.ts]: do not pass callbacks

File: setup/prompt.ts [hunk 9]
```diff
@@ -1,17 +1,23 @@
...
+        return "required";
       }
       return true;
     }
-  });
+  }).then((e: Partial<{ value: string }>) => e.value || "");
 
-  return sessionToken;
-};
+export const askSessionToken = () =>
```
Comment (gunzip) [inline setup/prompt.ts]: don't alias methods, use them directly instead

File: integrations/getServices/index.test.ts (hunk 2 above)
Comment (balanza) [reply integrations/getServices/index.test.ts]: Failing `expect`s actually throw. I don't know how to assert a code branch not to be executed (other than `expect(true).toBe(false)`).

Another solution would be implement `fold` callbacks as spies, and then check them to be called or not. But I think it's overkill other than very ugly.

Comment (balanza) [reply integrations/getServices/index.test.ts]: For the sake of code consistency, I may write a custom jest matcher that deals with a `fold`able structure. So we might have something like
```
expect(decoded).toBeRight()
```

Comment (gunzip) [reply integrations/getServices/index.test.ts]: https://jest-bot.github.io/jest/docs/expect.html#expectassertionsnumber

Comment (balanza) [reply integrations/getServices/index.test.ts]: anyway, I find this kind of failure annoying: 
Schermata 2020-03-05 alle 17 41 03

We might find a method to express such failure in a more comunicative way. Something like "you passed this, but it doesn't decode into your type"

Comment (gunzip) [reply integrations/getServices/index.test.ts]: I find it expressive (I read "expect the value to be a ServerInfo", what else ?)

File: integrations/getService/index.test.ts [hunk 10]
```diff
@@ -0,0 +1,86 @@
...
+    throw new Error(`required value dor "IO_BACKEND_HOST"`);
+  });
+  const basePath = getEnvValue("IO_BACKEND_BASEPATH").getOrElseL(() => {
+    throw new Error(`required value dor "IO_BACKEND_BASEPATH"`);
+  });
+  const sessionToken = getEnvValue("SPID_SESSION_TOKEN").getOrElseL(() => {
+    throw new Error(`required value dor "SPID_SESSION_TOKEN"`);
+  });
+  const fakeServiceId = "fake_id";
+  const existingServiceId = "io-dev-azure";
```
Comment (balanza) [inline integrations/getService/index.test.ts]: I hate this, but at the moment we have no other way to specify an existing service.

We could instead write a _scenario_  which fetches the list first, and them retrieve the single detail. But it seems a bit overkill to me.

Comment (gunzip) [reply integrations/getService/index.test.ts]: that's right, just use a different naming for mock objects and var:

`aFakeServiceId`
`anExistingServiceId`

so we can recognize it

File: lib/env.ts [hunk 11]
```diff
@@ -0,0 +1,43 @@
...
+  Object.entries(obj).reduce(
+    (p, [k, v]) =>
+      EnvMap.decode({ [k]: v }).fold(
+        () => p,
+        (o: EnvMap) => ({ ...p, ...o })
+      ),
+    {}
+  );
+
+const env: EnvMap = { ...pickEnvKeys(process.env) };
```
Comment (gunzip) [inline lib/env.ts]: it's very complex to make a code review of a PR that has dangling code. can you please do some clean up ?

**Comment by balanza** [comment]
I closed the PR as it's duplicated by #4

**Pull request closed without merging**